### ISSUE TITLE
RISC-V build enablement

### ISSUE THREAD

**Issue by Slackadays** [flathub/flathub#4378]
RISC-V is a new kind of CPU architecture that's free and open source, as well as pragmatically designed to be easy to learn yet powerful enough to be useful for real world projects. It's already well supported in the Linux kernel as well as most of userspace. With that, I think it's time that Flathub supported automatically building RISC-V packages alongside with amd64 and aarch64. 

I'm hoping this issue could become a tracker for RISC-V support in Flathub because there are a few things we need to consider along the way. These include:
- what hardware will Flathub use?
- where will the hardware be hosted?
- does Buildbot need any specific changes?
- does Flathub need any specific changes?

What does the Flathub team think of this? I'm willing to help out as much as I can with hardware, hosting, and software.

**Comment by doraskayo** [comment]
Freedesktop SDK essentially supports building its runtimes for RISC-V (specifically `riscv64`), but lacks hardware to build it. If you're not familiar with it, [Freedesktop SDK](https://gitlab.com/freedesktop-sdk/freedesktop-sdk) is the upstream runtime environment in which all of Flathub's apps and runtimes are built and run.

RISC-V is currently built in our x86_64 hardware (which is severely underpowered and overutilized) through emulation using `qemu-user-static`. It takes almost a week to build our runtimes from scratch for `riscv64`, compared to a few hours for every other architecture we support. As such, RISC-V can't be made a first-citizen in our CI and must be built manually once every few months. This also means that we don't test changes on RISC-V continuously, and would tag releases even if it fails to build.

In other words, without sufficient and sustainable hardware for CI we likely won't be able to include RISC-V in public releases.

With that said, we have everything other than hardware in place to supply `org.freedesktop.{Sdk,Platform}` for RISC-V, at least in theory.

**Comment by Slackadays** [comment]
If hardware is actually the only major blocker to RISC-V Flathub support, then that shouldn't actually be too bad. Right now Debian has a fleet of HiFive Unmatched boards which have about the same performance as a VisionFive 2 each, so it would be possible to do the same using those for a much lower cost. Alternatively, I happen to have preordered a Milk-V Pioneer which has 64 cores and 128GB of RAM, a much more formidable option. I'm willing to donate my hardware if that would solve this CI problem.

**Comment by nanonyme** [comment]
At this point following parties *at minimum* need riscv64 builders for it to work at all
1. freedesktop-sdk (base runtime and extensions)
2. Flathub (apps, KDE runtime)

Unless gnome also has riscv64 builders, then it's a no-go for any gnome riscv64 apps or runtimes.

**Comment by doraskayo** [comment]
@Slackadays, I think that Milk-V Pioneer-level hardware should be just about sufficient for Freedesktop SDK's CI needs, and could allow us to support RISC-V binaries as part of our official runtime releases.

However, while support in Freedesktop SDK would be the first major step, it's indeed not enough as @nanonyme mentioned. We'd also need hardware to provide CI services for Flathub's hosted apps (Buildbot), which I suspect may have even higher requirements if we wish to be able to build every app on Flathub for RISC-V. @barthalion, do you happen to know what kind of hardware would be required for this?

Perhaps RISC-V can be made available to apps on an opt-in basis until additional hardware is obtained. Anyway, a decision would need to be taken by the Flathub team.

There may also be additional infrastructure and website changes to support RISC-V on the Flathub side, but hopefully nothing too difficult.

CC: @jjardon

**Comment by Slackadays** [comment]
I'm pleased to announce I finally got a Milk-V Pioneer from the RISC-V Foundation specifically for this issue. I've already tried getting the Freedesktop SDK running but I keep getting into dependency hell with BuildGrid and Buildstream, which doesn't seem to be specific to RISC-V. Do we know yet what hardware we actually need for the buildbot?

**Comment by Slackadays** [comment]
@nanonyme What should I do or who can I talk to once the qemu-binfmt image is built?

**Comment by nanonyme** [comment]
If you have native image, you can consume it as toolbox and use it to build runtimes. It contains sufficient tooling for building everything. For riscv64 to actually be supported, freedesktop-sdk, gnome and Flathub all need performant physical riscv64 hardware. If there are willing sponsors, @barthalion and @jjardon are probably good points of contact.

**Comment by jjardon** [comment]
We have another pioneer for the freedesktop-sdk project (which I hope to setup in the coming weeks); @Slackadays I would recommend to use the one you have to help with the flathub infrastructure (ie, apps); best contact person for that is indeed @barthalion

**Comment by nanonyme** [comment]
registry.gitlab.com/freedesktop-sdk/infrastructure/freedesktop-sdk-docker-images/bst2 manifest aims to have riscv64 compatible manifest. It can be consumed as toolbox as well as CI image. Latest manifest where riscv64 is guaranteed to be present and functioning correctly is registry.gitlab.com/freedesktop-sdk/infrastructure/freedesktop-sdk-docker-images:272055e8a884e18adc156ce1c26cd279e4aa8609.

**Comment by Slackadays** [comment]
@barthalion Did you see my email?

**Comment by Haui1112** [comment]
hello folks! I'm a huge fan of both flatpak and riscv. for that reason I bought a DC-ROMA PAD II with Spacemit K1 8 RISC-V Cores and 16 GB of ram. Please let me know if you need stuff compiled. Thats exactly why I bought this.

**Comment by Slackadays** [comment]
I am still waiting for any opportunity to use my Milk-V Pioneer which has been gathering dust in a corner for over a month

**Comment by jjardon** [comment]
> I am still waiting for any opportunity to use my Milk-V Pioneer which has been gathering dust in a corner for over a month

@Slackadays If flathub is not ready yet for this, happy to use your machine in freedesktop-sdk (https://freedesktop-sdk.io/), which is the base for all the runtimes. Please PM me in matrix and we can get set it up quickly (@jjardon:matrix.org), or email me if you prefer (jjardon@gnome.org)

**Comment by Slackadays** [comment]
Update for onlookers: We are now working to build freedesktop-sdk

**Comment by Slackadays** [comment]
> That is SO AWESOME! Thank you for helping out!

If you want to help with your ROMA PAD too, check out https://udd.debian.org/cgi-bin/ftbfs.cgi?arch=riscv64 for a list of things that need RISC-V porting!

**Comment by archanox** [comment]
Is there any updates on what the progress is for the RISC-V builders?
What are the blockers? Missing hardware? Unported software dependencies?

Am I able to donate non-dedicated, unpredictable uptime compute for a builder with my Pioneer?

**Comment by nanonyme** [comment]
Porting is more or less fine. Base runtime (freedesktop-sdk) builds but build performance even with fastest sponsored riscv64 is on completely different ballpark as x86_64, aarch64 and ppc64le. @jjardon might be able to share some more granular information. At current performance the riscv64 runners freedesktop-sdk has cannot keep up with the level of updates landed in freedesktop-sdk. We could possibly provide freedesktop-sdk riscv64 through freedesktop-sdk internal flatpak repo for enthusiasts but Flathub seems still reasonably far in the horizon.

**Comment by Slackadays** [comment]
Another issue is that there's going to be no more new Pioneers in the foreseeable future because of the new Sophgo sanctions. However, there's more, faster hardware on the horizon, so we'll have to see if that can help us here. I also think an "enthusiast" repo for Flathub would be a fine option like @nanonyme said.

**Comment by Vixea** [comment]
Any updates on this "enthusiast" repo I'm looking at adding RISC-V support to the flatpak version of prism-launcher? I can explain things in more detail if needed.

**Comment by bbhtt** [comment]
No immediate plans to add RISC-V support.
